util/resolver/resolver.go

```diff
@@ -236,23 +238,41 @@ func hostsWithCredentials(hosts docker.RegistryHosts, auth *SessionAuthenticator
 	if hosts == nil {
 		return nil
 	}
+	cache := map[string][]docker.RegistryHost{}
+	var mu sync.Mutex
+	var g flightcontrol.Group
 	return func(domain string) ([]docker.RegistryHost, error) {
-		res, err := hosts(domain)
+		v, err := g.Do(context.TODO(), domain, func(ctx context.Context) (interface{}, error) {
+			mu.Lock()
+			v, ok := cache[domain]
+			mu.Unlock()
+			if ok {
+				return v, nil
+			}
+			res, err := hosts(domain)
+			if err != nil {
+				return nil, err
+			}
+			if len(res) == 0 {
+				return nil, nil
+			}
+
+			a := docker.NewDockerAuthorizer(
+				docker.WithAuthClient(res[0].Client),
+				docker.WithAuthCreds(auth.credentials),
+			)
+			for i := range res {
+				res[i].Authorizer = a
+			}
+			mu.Lock()
+			cache[domain] = res
+			mu.Unlock()
+			return res, nil
+		})
 		if err != nil {
 			return nil, err
 		}
-		if len(res) == 0 {
-			return nil, nil
-		}
-
-		a := docker.NewDockerAuthorizer(
-			docker.WithAuthClient(res[0].Client),
-			docker.WithAuthCreds(auth.credentials),
-		)
-		for i := range res {
-			res[i].Authorizer = a
-		}
-		return res, nil
+		return v.([]docker.RegistryHost), nil
```

Comment:
If L256 is return nil, the assertion will panic here?

Reply:
Good catch. Did not appear in practice because `hosts` will never return `nil` in buildkitd.